### PR TITLE
Added an options menu

### PR DESCRIPTION
**Pull request by NancokPS2**
Includes a volume slider (for the main and currently only bus) and button remapping.  
game_menu.gd got a small rework to better fit sub-menus like the new options_menu.tscn scene.  
  
Button remapping works by listing action names and it's events to the right as buttons, clicking a button prepares the system for a remapping attempt and the next valid button will replace the event.   
Currently the actions appear raw, setting up a translation file could easily solve this. (for another PR)
  
There's currently no serialization and therefore no permanent save of options (but again, that's for another PR).

### CODE REVIEW

**Comment by NancokPS2** [comment]
Closes #41 ?

**Comment by jonathaneeckhout** [comment]
Another great PR, thanks.

Some feedback regarding the menu:

- The option tab text is very small on the left top corner, the words: Button mapping, audio could be bigger.
- The menu has no margin from the sides. The text immediately starts on pixel 1 of my left side of the screen. I think it would be nicer to make the menu have margin or make the menu smaller.
- The remapping works, but I think it should show some feedback while mapping. If you click a mapping, it should say something like: "Press and key to map or Esc to cancel".

Screenshot from 2023-10-12 13-59-01

**Comment by NancokPS2** [comment]
Will work on these.  
Altho the bag thing may be easier to solve by setting up translations instead of making a hardcoded string change.  
  
I'm already doing a PR for adding a translation file.

File: scenes/ui/options_menu/input_remapping.gd [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+var currentAction:StringName
+var currentEvent:InputEvent
+var currentButton:Button
+
+func _ready() -> void:
+	update_list()
+	
+func update_list():
+	#Clear all children
+	for child in get_children(): child.queue_free
```
Comment (jonathaneeckhout) [inline scenes/ui/options_menu/input_remapping.gd:27]: I think you forgot the "()" for queue_free().
The child.queue_free() should also be placed on the next line

File: scenes/ui/options_menu/input_remapping.gd [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+class_name InputRemappingUI
+##Recommended for use in a VBoxContainer
+##Fills itself with buttons to remap actions
+
+
+const BUTTON_ACTIVE_MODULATE:Color = Color.WHITE * 1.5
+
+#This needs tweaking
+@export var actionsAllowed:Array[StringName] = [
+#	&"j_left_click",
```
Comment (jonathaneeckhout) [inline scenes/ui/options_menu/input_remapping.gd:11]: Remove commented code

File: scenes/ui/options_menu/input_remapping.gd [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	currentEvent = null
+	currentButton = null
+	
+	
+func _input(event:InputEvent):
+	#If these are not set, abort
+	if currentAction.is_empty(): return
+	if currentEvent == null: return
+	if currentButton == null: return
+#	print_debug("Can remap")
```
Comment (jonathaneeckhout) [inline scenes/ui/options_menu/input_remapping.gd:92]: Remove commented code

**Comment by NancokPS2** [comment]
> Another great PR, thanks.
> 
> Some feedback regarding the menu:
> 
>     * The option tab text is very small on the left top corner, the words: Button mapping, audio could be bigger.
> 
>     * The menu has no margin from the sides. The text immediately starts on pixel 1 of my left side of the screen. I think it would be nicer to make the menu have margin or make the menu smaller.
> 
>     * The remapping works, but I think it should show some feedback while mapping. If you click a mapping, it should say something like: "Press and key to map or Esc to cancel".
> 
>     * The j_toggle_bag is only meaningful for the game, a user should get the text: Toggle your bag
> 
>     * The slider of the audio volume could be bigger, I didn't notice it in the beginning (it kinda blended with one of the tents in the background).
> 
> 
> Screenshot from 2023-10-12 13-59-01
> 
>     * When you're mapping a mapping the key is actually used by the game. If I remap the toggle bag, the bag actually toggles. I think this can be solved by setting the the above_ui variable.
> 
>     * Whenever I remap a key I get this error on client side the terminal:
> 
> 
> Screenshot from 2023-10-12 14-16-05
  
- The slider size is tied to the theme i believe. I am a bad artist, altho i did make it actually fit with the size of the text instead of cramping into the top of it's container (this may make it functionally bigger too, i cannot really tell)  
   
- The issue with the bag opening is on the main branch as well, having the pause menu open still allows opening UI elements like thos.   
- JUI.above_ui does not actually stop any inputs from reaching the player. I am unsure of how to fix this without reworking node hierarchies and whatnot. Which may not be the best to include in this PR. Maybe using _unhandled_input in some places could help.    
- The last error was because the action had a leading space, fixed now.  }

**Comment by jonathaneeckhout** [comment]
Can you rebase this one now https://github.com/jonathaneeckhout/jdungeon/pull/65 has been merged?

**Comment by jonathaneeckhout** [comment]
Tested the latest version and the looks and feels are much better.

However I still encountered a bug:
when you map the same key twice, so like use the "h" key for bag and equipment and remap it again afterwards. Toggle the menu a couple of times you can get into a state that you have double mappings:
Screenshot from 2023-10-13 15-29-18

I think there should be a check before the mapping to double check that the mapping isn't being used somewhere else.

Lastly I have this remark:
The option menu does not have a "close" button.

The only way to close it, is to press esc which will close the whole menu. For a smoother user experience I thing adding a close button would be nice.

**Pull request merged**